src/Sentry/Internal/CachingBackgroundWorker.cs

```diff
@@ -0,0 +1,146 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Sentry.Extensibility;
+using Sentry.Protocol.Envelopes;
+
+namespace Sentry.Internal
+{
+    internal class CachingBackgroundWorker : BackgroundWorkerBase
+    {
+        private const string EnvelopeFileExt = "envelope";
+        private readonly DirectoryInfo _cacheDirectory;
+
+        public CachingBackgroundWorker(ITransport transport, SentryOptions options)
+            : base(transport, options)
+        {
+            _cacheDirectory = !string.IsNullOrWhiteSpace(options.CacheDirectoryPath)
+                ? _cacheDirectory = new DirectoryInfo(options.CacheDirectoryPath)
+                : throw new InvalidOperationException("Cache directory is not set.");
+        }
+
+        private IEnumerable<FileInfo> GetEnvelopeFiles() =>
+            _cacheDirectory.Exists
+                ? _cacheDirectory.EnumerateFiles($"*.{EnvelopeFileExt}")
+                : Enumerable.Empty<FileInfo>();
+
+        private FileInfo? TryGetNextEnvelopeFile() =>
+            GetEnvelopeFiles()
+                .OrderBy(f => f.CreationTimeUtc)
+                .FirstOrDefault();
+
+        private async ValueTask FlushCacheAsync(
+            CancellationToken cancellationToken = default)
+        {
+            Options.DiagnosticLogger?.LogDebug("Flushing cached envelopes.");
+
+            while (TryGetNextEnvelopeFile() is { } envelopeFile)
+            {
+                Options.DiagnosticLogger?.LogDebug(
+                    "Reading cached envelope: {0}",
+                    envelopeFile.FullName
+                );
+
+                using var envelope = await Envelope.DeserializeAsync(
+                    envelopeFile.OpenRead(),
+                    cancellationToken
+                ).ConfigureAwait(false);
+
+                Options.DiagnosticLogger?.LogDebug(
+                    "Sending cached envelope: {0}",
+                    envelope.TryGetEventId()
+                );
+
+                try
+                {
+                    await Transport.SendEnvelopeAsync(envelope, cancellationToken).ConfigureAwait(false);
+
+                    // Delete the cache file in case of success
+                    envelopeFile.Delete();
+
+                    Options.DiagnosticLogger?.LogDebug(
+                        "Successfully sent cached envelope: {0}",
+                        envelope.TryGetEventId()
+                    );
+                }
+                catch (IOException ex)
```

Comment:
I'm afraid we could have some I/O error that's not really recoverable and we'll be retrying that forever. Maybe we need to be more strict on what we consider OK to retry

Reply:
Unfortunately most transient errors we would want to retry on just throw base `IOException`. Some have inner exceptions (like `SocketException`) but not all and it's probably implementation-dependent. I guess we'd have to parse exception messages if we really wanted to be selective.

Reply:
That's an option. For now we're good with IOException, the client returning 500 wouldn't throw so at least we know that if it hits Sentry it'll be deleted